mm2src/mm2_io/src/fs.rs

```diff
@@ -276,10 +271,88 @@ where
     read_files_with_extension(dir_path, "json").await
 }
 
+/// Creates all the directories along the path to a file if they do not exist.
+pub fn create_parents(path: &dyn AsRef<Path>) -> IoResult<()> {
+    let parent_dir = path.as_ref().parent();
+    let Some(parent_dir) = parent_dir else {
+        return MmError::err(
+            io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("{} has no parent directory", path.as_ref().display()),
+            ))
+    };
+    match fs::metadata(parent_dir) {
+        // Path exists, make sure it's a directory (and not a file for example).
+        Ok(metadata) => {
+            if !metadata.is_dir() {
+                return MmError::err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("{} is not a directory", parent_dir.display()),
+                ));
+            }
+        },
+        // This path doesn't exist, create it.
+        Err(_) => fs::create_dir_all(parent_dir)?,
+    }
+    Ok(())
+}
+
+/// Similar to [`create_parents`], but using non-blocking async IO operations.
+///
+/// Creates all the directories along the path to a file if they do not exist.
+pub async fn create_parents_async(path: &Path) -> IoResult<()> {
+    let parent_dir = path.parent();
+    let Some(parent_dir) = parent_dir else {
+        return MmError::err(
+            io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("{} has no parent directory", path.display()),
+            ))
+    };
+    match async_fs::metadata(parent_dir).await {
+        // Path exists, make sure it's a directory (and not a file, for instance).
+        Ok(metadata) => {
+            if !metadata.is_dir() {
+                return MmError::err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("{} is not a directory", parent_dir.display()),
+                ));
+            }
+        },
+        // This path doesn't exist, try to create it.
+        Err(_) => async_fs::create_dir_all(parent_dir).await?,
+    }
+    Ok(())
+}
+
+/// Writes the `content` to the file at `path`.
+///
+/// This also creates any intermediary directories up to the file itself if they do not exist.
+/// If `use_tmp_file` is true, it writes to a temporary file first and then renames it to the final file name
+/// to ensure atomicity.
+pub fn write(path: &dyn AsRef<Path>, content: &[u8], use_tmp_file: bool) -> IoResult<()> {
```

Comment:
Why do we need dyn here,
isn't it enough just `impl AsRef<Path>` here in the write fn and in create_parents fn?

Reply:
dunno why i went this way 😅 , prolly cuz i like dynamic dispatch.

reverting to `impl AsRef` for consistency with the codebase. [e1f478f](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2398/commits/e1f478feb0693fd3d3094c9f7b76981a894fd172)